Issue #181 Hotbar Item Order Persistence

rushab101 opened the pull request:
The following PR contains an implementation for the Hotbar to be persisted split into several sections.

### Part A: Establishing the localStorage

Since an end-user can have multiple accounts, simply store the hotbar for one account is a flawed design choice. User A has two accounts: AccountA, and AccountB, each with its own hotbar. User A may want to log into Account A, then log out, and then log into AccountB, so when saving into localStorage there must be a way to uniquely identify which hotbar to load into which account. To do this, when loading the hotbar right after the user logged in, I passed into two additional parameters from: \dungeonz\client\src\game\Game.js which are checking if a user is logged in, and the userName of the user. I confirmed that each account has a unique userName so this can work to uniquely identify which account it is. If the user is not logged in, the hotbar cannot be saved and thus the old logic of the hotbar is followed. 

### Part B: Making hotbar persistence
**Initial Solution:** 
As stated in the issue, each item has a unique ID so initially the problem seemed to be very trivial. We can use a set data structure, and now we have to account for three cases:

Case A: Adding into hotbar: 
When an item has been added to the hotbar, add the unique ID of that item into the set, and then save the set to localStorage

Case B: Removing from hotbar:
When an item has been removed from the hotbar, remove the unique ID of that element from the set, and then update the localStorage

Case C: Loading the hotbar
When the game is loaded, if the user is not logged in follow the old logic of populating the hotbar. If the user is logged in, and the localStorage does not exist, follow the old logic of populating the hotbar. If the user is logged in and the localStorage exists, load the set from the localStorage. Next, iterate through each item in the inventory, and if the item id from inventory matches the item id in the set, add it to the hotbar.

Problems with this solution:
It may be that I missed something, but each time I log in, into the same account by refreshing the browser, the unique item ID changes. It seems to be randomized each time, and so a match from the set to the inventory will never be the case and this solution would not work. Images below demonstrate the issue:

User ‘bbb’ logs in:
![image](https://user-images.githubusercontent.com/43216310/144979560-0fdddc47-7f02-487b-8e9f-c30081de2fe7.png)
User ‘bbb’ refreshes the page and then relogs in:
![image](https://user-images.githubusercontent.com/43216310/144979603-22cb4a0b-7af8-4413-8435-d17f83660bc8.png)
We can see with these two screenshots that without modifying anything, and simply refreshing the page the unique IDs are randomly generated, and you cannot make something randomly generated each time to be persistent. 

**Alternative Solution:**
The criteria stated that the ordering of items in the hotbar must be saved, and the operations of adding, removing and loading must be persistent. Therefore since we guarantee that each item in the hotbar must be loaded from inventory we can use a hashmap, and do a 1-to-1 mapping from inventory to hotbar and save the hashmap into the localStorage. To do this, we know that each item has a field called inventory slot and each slot in the hotbar is also a slot, so it makes the mapping possible. This does introduce several more cases to be concerned about which I will address individually below.

Case A: Adding items from inventory to hotbar
This is the simplest case. Since we are appending the item to the end of the hotbar, this would be simply: hashmap[hotbar.lenght-1= itemConfig.slotIndex. Then we simply save the hashmap to local storage





Case B: Removing items from hotbar
This is a lot more difficult because the user can remove an element from any part of the hotbar, and we have to ensure that both the slotIndex and the hotbar Index is always updated or else the entire implementation fails. Due to the several edge cases in this scenario I have broken it down into several sub components:

Case B-1: Removing Element from hotbar
With hashmaps, you remove entries by the keys. In this case: key = hotbar Index, and value = Inventory Index. Therefore when removing the element from the hotbar, find the corresponding key that is related to the inventory Index, and that is the key we remove from our hashmap. However, since an element from the hotbar is removed, you have to adjust all the other indexes that come after that key and subtract it by 1, or else we would not get the correct index. The following images showcases this:

Use Case: User ‘bbb’ decides to remove the second item (Potion) from the hotbar. 

Before: 
![image](https://user-images.githubusercontent.com/43216310/144979865-c8c73f26-d8f5-4cf3-9c1d-ed4fbe76d3b9.png)
How HashMap Looks:
![image](https://user-images.githubusercontent.com/43216310/144979773-2c4eff9b-1cb9-453a-be16-dd96f5dd3279.png)
After removing the potion and refreshing the page
![image](https://user-images.githubusercontent.com/43216310/144979809-0a50821a-0470-4cbf-a564-536ff42761de.png)
How HashMap Looks:
![image](https://user-images.githubusercontent.com/43216310/144979936-7f373a28-6b1b-4878-bb88-1d5cd155a5d3.png)
Case B-2.1: Removing Element from inventory
This case assumes that the inventory index is found within the hashmap. When we remove an item from inventory we have to update both the hotbar index and inventory index. Therefore the implementation will follow the previous logic with the addition of reducing the inventory index by 1 starting from the key which was removed from the hashmap. The following images below showcases this: 
Use Case: User ‘bbb’ decides to drop the 4th  item in its inventory : 4 windstaff 
Before: 
![image](https://user-images.githubusercontent.com/43216310/144980013-29de8584-c2fe-499b-9bc0-eb408ea5bc60.png)
How HashMap Looks:
![image](https://user-images.githubusercontent.com/43216310/144980122-eefaf7d3-c62e-4fbc-92d4-2a3974b4d5e7.png)
After dropping the windstaff, refreshing the page and logging back in
![image](https://user-images.githubusercontent.com/43216310/144980191-7267053e-1e3a-43fe-884c-470cf8b8ceaf.png)
How HashMap Looks:
![image](https://user-images.githubusercontent.com/43216310/144980231-8e9b1540-9585-4b3d-891a-71d085bc30ac.png)

Case B-2.2: Removing Element from inventory where the slotIndex does not exist in the hashmap
Previously to find the key to delete from hashmap, I was using this line: Object.keys(this.mapKey).find((key) => this.mapKey[key] === itemConfig.slotIndex);
However, if an item is deleted from the inventory and was not added in the hashmap (the end user never added this item to hotbar or the item cannot be added to the hotbar), this will return null. In this case, only the slotIndex were modified but we still have to update our hashmap values to represent this change. Therefore we use the slotIndex that was passed into the remove function, and iterate through the hashmap, and using this we can find out which slotIndex we have to decrease by 1 in our hashmap. The following below demonstrates this case:

Use Case: User, ‘bbb’ adds a bunch of items, maxes out on inventory, and needs to drop some items to clear up space, some of which items, have never been added to the hotbar

Before: 
In this Case the items that have square border, element represents items that has never been added to hotbar, and the X represents the elements we will drop
![image](https://user-images.githubusercontent.com/43216310/144980336-9efe3f41-d5e0-46ac-b96e-1a49950b9977.png)
How Hasmap looks:
![image](https://user-images.githubusercontent.com/43216310/144980384-0d8a3e4d-30d0-4a71-a3db-025e5ef686f2.png)
Refresh the page and log back in
![image](https://user-images.githubusercontent.com/43216310/144980449-50df99de-dc6f-4de6-a1e9-87bd5a45e194.png)
How Hasmap looks like
![image](https://user-images.githubusercontent.com/43216310/144980508-4d652a6b-39e8-4695-a57b-4c12f0a78812.png)
Case C: Loading the hotbar after user logs in
First check if the user is logged in, if the user is not , follow the original method of initializing the hotbar. If the user is logged in , and the localStorage does not exist , follow the original method. If the localStorage is found for the account, load the hashmap from the localStorage. Before, using the data from hashmap, first we have to do some sanity checking to ensure that any corruptions would not cause the server to crash. Right now, if the size of the hashmap is 0, it will follow the original method of initializing the hotbar → (User empties out the hotbar, and then re-logs back in). This can be changed if not wanted. Next we do the following sanity checks:

1. Checking if multiple keys hold the same data → Incorrect Mapping
2. Checking if any indexes is less than 0 → Hash map has not been updated properly after a remove operation or there is a localStorage mismatch
3. Checking if any keys are greater than the previously saved hotbar length →  Hash map has not been updated properly after a remove operation or there is a localStorage mismatch
4. Checking if any values are greater than the size of the inventory →  Hash map has not been updated properly after a remove operation or there is a localStorage mismatch
5. Checking if there is a localStorage mismatch


If any conditions triggered: A a warning appears, using Utils.warning(), and this implies there is some logic error somewhere, and missed an additional edge case. After the warning, we clear the hashmap, and follow the original method of initializing the hotbar and then save this new hotbar into localStorage.

No conditions triggered: This means everything was successfully loading, and now we populate the indexes of the hotbar, with the corresponding index from the inventory using the hashmap. 

### Part C: Validation and Quality Assurance
After the implementation was complete, I created a checklist and validated each option via manual quality assurance.
![image](https://user-images.githubusercontent.com/43216310/144980738-ad913567-3f4f-428b-88b7-a0227689e305.png)
![image](https://user-images.githubusercontent.com/43216310/144980794-161e79ec-c94e-44a1-9650-f7ad63e31091.png)
![image](https://user-images.githubusercontent.com/43216310/144981177-0ea49815-8f96-48e4-87ac-197047cfaced.png)

### Part D: Moving Forward
I would say that the solution provided within this PR is a temporary workout and moving forward I would suggest two two other options for more permanent solution to this issue:

Option A: Create a new issue, and somehow get the item IDs to be persistent meaning that after logging in, logging back out, and then logging back in, the IDs should not change for the existing items in the inventory. Then it heavily simplifies the implementation and if needed I can easily modify the code and implement the original way it was intended. Using the ID match method also removes almost all the sanity checks in the existing solutions.  

Option B: If for whatever reason the item IDs cannot be persistent, then to store the hashmap directly into the MongoDB database instead of using localStorage. This method will completely remove any localStorage mismatch. Furthermore, as an end user of the game I would rather have it so that if I log in first time from my computer into the game, and then log in first time from my laptop, the hotbar should be synced. This is how the majority of other MMOs have their hotbars implemented.

Please note I am not confident with this alternative design and feel like I missed some additional test cases which could cause the mapping to become incorrect. Further playtesting needed.

rushab101 committed:
Temp Solution

rushab101 committed:
Done Second Draft for PR

rushab101 committed:
Fixed up some bugs

rushab101 committed:
Cleaned Up Code

rushab101 committed:
Changed back volume to default

gabriellemadden commented on client/src/shared/state/Inventory.js:
```diff
@@ -104,16 +107,155 @@ class Inventory {
+        window.localStorage.removeItem(localStorageID);
+        window.localStorage.setItem(localStorageID, JSON.stringify(this.keyToSlotIndex));
+        // Used to debug
+        // Utils.message(this.keyToSlotIndex);
```
remove comments if you don't need them anymore

gabriellemadden commented on client/src/shared/state/Inventory.js:
```diff
@@ -104,16 +107,155 @@ class Inventory {
+
+    saveHotbar() {
+        const sessionIDMap = "saveHotBar";
+        const localStorageID = this.userName + sessionIDMap;
```
is this a standard way to store stuff in local storage? You might want to include some way to uniquely identify that the user is coming from dungeonz (incase they use the same username for other sites). I would ask the owners of the repo what they'd like to use.

rushab101 replied:
Even if they did use the userName for another websites, the chance that it's gonna be "bbb" + "saveHotBar" --> "bbbsaveHotBar" is soo small, but if you want I can hash it, and store the hash as the unique idetifer.

gabriellemadden replied:
identical keys clashing isn't really the concern, it's more for if someone were to inspect their localstorage and they see "bbbsaveHotBar" then there's a chance they wouldn't know which website it came from. It would be better to have some way to identify the origin, like `dungeonz_bbb_saveHotBar` (then for other localstorage items dungeonz would follow the convention `dungeonz_bbb_blahblah`). But, the owners of dungeonz would have a better idea than me on what this convention would be (or is, if they already have established it)

rushab101 replied:
Or alternatively I can concatenate userName + (16 bit hash) --> Gaurantee to be unique regardless of what website

rushab101 replied:
![image](https://user-images.githubusercontent.com/43216310/145159369-9a1b4f4e-022b-49aa-bb7e-42fe165db164.png)
This is an example of how they did it.

gabriellemadden replied:
I see, that's bad key use on them then I guess 🤷🏼‍♀️

gabriellemadden commented on client/src/shared/state/Inventory.js:
```diff
@@ -104,16 +107,155 @@ class Inventory {
+        });
+    }
+
+    removeFromHotbar(itemConfig, removedInventory = false) {
```
booleans in function parameters are a code smell: https://understandlegacycode.com/blog/what-is-wrong-with-boolean-parameters/

You might want to break this function apart into smaller functions that perform the necessary isolated behaviours.

What does it mean when you call this function with `removedInventory=true`, does it mean the player dropped an item from their inventory and so consequentially the item must also be removed from the hotbar?

rushab101 replied:
Yes, pretty much, removed from inventory so, update both the key and value, instead of updating just the key, I will see if time permits, to break it up.

rushab101 replied:
Yeah, from my through playtesting I have not encountered an issue, for now I am gonna leave it.

gabriellemadden commented on client/src/shared/state/Inventory.js:
```diff
@@ -104,16 +107,155 @@ class Inventory {
+
+    loadHotbar(playerID, isLoggedIn) {
+        if (isLoggedIn === true) {
+            this.saveSession = true;
```
what would happen if `saveSession` were true if the user is not logged in? (i know the code would never reach this point as it is, just curious if it's necessary)

rushab101 replied:
It will set the username == null, and then ur doing the calling the local storage --> key = (Null)

gabriellemadden replied:
it might be better to check if the username is null every time you do a localStorage operation rather than have a separate state saved in the class, to make the code more readable. What do you think?

gabriellemadden commented on client/src/shared/state/Inventory.js:
```diff
@@ -104,16 +107,155 @@ class Inventory {
+    }
+
+    saveHotbar() {
+        const sessionIDMap = "saveHotBar";
```
You have duplicate code (see line 215) - this means this constant should be saved as an `Inventory` class field so you can access it from all class functions. This would make it easier to change the sessionIDMap string later if they wanted to. Also, just nitpicking, but I would probably use a more descriptive variable name rather than `sessionIDMap` so people reading the code know what its purpose is. Maybe something like `HotBarStorageKey`

gabriellemadden commented on client/src/shared/state/Inventory.js:
```diff
@@ -104,16 +107,155 @@ class Inventory {
+        }
+        this.userName = playerID;
+        const sessionIDMap = "saveHotBar";
+        const localStorageID = this.userName + sessionIDMap;
```
if you move `sessionIDMap` into a class field, then `localStorageID` could be changed to a class method also (moved to the class definition at top of file)
```suggestion
        function localStorageID() {return this.userName + this.sessionIDMap};
```

gabriellemadden commented on client/src/shared/state/Inventory.js:
```diff
@@ -104,16 +107,155 @@ class Inventory {
+            } else if (Object.keys(this.keyToSlotIndex).length !== 0) {
+                this.hotbar[key] = this.items[this.keyToSlotIndex[key]];
+                PubSub.publish(HOTBAR_ITEM);
+                // PubSub.publish(MODIFY_INVENTORY_ITEM);
```
```suggestion
```

gabriellemadden commented on client/src/shared/state/GUI.js:
```diff
@@ -59,14 +59,14 @@ class GUI {
          * @type {Number}
          */
-        this.musicVolume = 0;
+        this.musicVolume = 50;
```
😆  I would probably amend the commit that you initially changed this in so that there is no history of someone changing the volume, to keep the master history clean

gabriellemadden replied:
Looks like upstream squashes commits when they merge anyway, so this might not matter

gabriellemadden reviewed (commented):
It's interesting that all 3 package-lock.json files changed so much without actually updating any dependencies in any package.json. Upstream might not want to include all the dependency changes (could break other features). The standard procedure might just be to not commit them (if that's the case though, they should probably have been added to  .gitignore by the owners...) I would include all the package.lock changes as a separate commit just in case they don't want to merge.

Besides that, just a few styling changes I would do, but it looks good otherwise. I think the best thing to do before making major changes is see what upstream thinks of your workaround/if they have any advice. PR summary is super descriptive, very easy to read, big 👍🏼  for that

rushab101 committed:
Cleaned Up Code + added Comments

rushab101 committed:
Reverted lock.json

rushab101 committed:
Reverted all package-lock JSON files

rushab101 committed:
Final Clean Up

rushab101 committed:
Changed playerName to playerID

PlayerName was not unique, so changed it to playerID to properly represent this

Arcanorum committed:
Make item config ids persistent on accounts

Arcanorum committed:
Add support for item ids on older documents

rushab101 committed:
Merge remote-tracking branch 'upstream/master'

rushab101 committed:
Implemented Method 1 in PR

rushab101 committed:
Added Required Changes from PR
